pyodide-build/pyodide_build/buildall.py

```diff
@@ -242,6 +251,23 @@ def get_progress_line(package_set):
     return f"In progress: " + ", ".join(package_set.keys())
 
 
+def format_name_list(l):
+    if len(l) == 1:
+        return l
+    most = l[:-1]
+    if len(most) > 1:
+        most = [x + "," for x in most]
+    return " ".join(most) + " and " + l[-1]
+
+
+def mark_package_needs_build(
+    pkg_map: Dict[str, BasePackage], pkg: BasePackage, needs_build: Set[str]
+):
+    needs_build.add(pkg.name)
+    for dep in pkg.dependents:
+        mark_package_needs_build(pkg_map, pkg_map[dep], needs_build)
+
```

Comment:
Maybe better to return the needs_build and aggregate the output explicitly, rather than having a recursive function with side effects on  input variables,

```suggestion
    pkg_map: Dict[str, BasePackage], pkg: BasePackage
) -> Set[str]:
    needs_build = set(pkg.name)
    for dep in pkg.dependents:
        needs_build.update(mark_package_needs_build(pkg_map, pkg_map[dep]))
    return needs_build

```

Reply:
This is basically the same implementation as `addPackageToLoad`:
https://github.com/pyodide/pyodide/blob/42180e86816e4db176d4bb770f0a897c600d7438/src/js/load-pyodide.js#L111
I'll try adding a separate function which does the base case of the recursion.